File: ruby/palindrome-products/palindrome_products.rb
# frozen_string_literal: true

class Palindromes
  # Palindrome = Structure.new(:value)
  attr_reader :largest, :smallest

  def initialize(input)
    @max_factor = input[:max_factor]
    @min_factor = input[:min_factor] || 1
    @range = (@min_factor..@max_factor)
    @palindromes = palindromes
  end

  def generate
    @largest = Palindrome.new(@palindromes.max, factors(@palindromes.max))
    @smallest = Palindrome.new(@palindromes.min, factors(@palindromes.min))
  end

  def palindromes
    array = []
    @range.map do |num|
      @range.map do |other|
        factor = other * num
        array << factor if factor.digits.reverse == factor.digits
      end
    end
    array
  end

  def factors(number)
    numbers = @range.select do |num|
      (number % num).zero? && @range.include?(number / num)
    end
    numbers.map { |num| [num, number / num].sort }.uniq
  end
end

class Palindrome
  attr_reader :value, :factors
  def initialize(value, factors)
    @value = value
    @factors = factors
  end
end

File: javascript/palindrome-products/palindrome-products.js
export class Palindromes {
  static generate(parameter) {
    let minFactor = parameter.minFactor;
    let maxFactor = parameter.maxFactor;

    let isPalindrome = product => {
      return (
        Number(
          String(product)
            .split("")
            .reverse()
            .join("")
        ) == product
      );
    };

    let palindromes = [];

    for (let i = minFactor; i <= maxFactor; i++) {
      for (let j = minFactor; j <= maxFactor; j++) {
        let product = i * j;

        if (isPalindrome(product)) {
          /*console.log()*/
          if (!palindromes[product]) {
            palindromes.push({ value: product, factors: [i, j] });
            console.log(product);
          }
          /* if  (
              !palindromes[product].factor.includes(i) ||
              !palindromes[product].factor.includes(j)
            ) {
              palindromes[product].factor.push([i, j]);
            } */

          /*  
          } */

          /*  console.log(product); */
          /* if (!palindromes[product]) {
            palindromes.product = [];
          }

          */
        }
      }
    }
    console.log(palindromes);
    /* while (currentValue < numbers.maxFactor) {
      if (i < numbers.maxFactor) {
        i++;
      } else {
        currentValue++;
      }
      product = currentValue * i;
      console.log(currentValue, i, product); */
    /* if (
        Number(
          String(product)
            .split("")
            .join("")
        ) ==
        Number(
          String(product)
            .split("")
            .reverse()
            .join("")
        )
      ) {
        palindromeProducts.push({ value: product, factors: [currentValue, i] });
      }
      console.log(palindromeProducts);
      console.log(palindromeProducts[0].value); */
    /* return {
        smallest: {
          value: palindromeProducts[0].value,
          factors: [palindromeProducts[0].factors]
        },
        largest: palindromeProducts[palindromeProducts.lastIndexOf() - 1]
      }; */
    /* } */
  }
}
